---
title: "Exploring CompHeatmaps"
output: rmarkdown::html_vignette
author: "Koji Wong"
vignette: >
  %\VignetteIndexEntry{CompHeatmaps}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

To download **CompHeatmaps** use the following commands
```{r}
require("devtools")
devtools::install_github("kojiwong/CompHeatmaps", build_vignettes = TRUE)
# load the library
library("CompHeatmaps")
```

First we want to get our data into the right format.
```{r}
# Format our data to put into the functions
current_dir <- getwd()
# Make sure that current dir is mapped to package folder and not vignettes folder.
print(current_dir)
sample_data_path <- file.path(current_dir, "../data/inst/extdata/sample_raw_16S_data")
sample_output_path <- file.path(current_dir, "../data/inst/extdata/filtered_reads")
sample_data_path
list.files(sample_data_path)
sample_list <-sort(list.files(sample_data_path, pattern=".fastq.gz", full.names = TRUE))
sample_list
sample_data.names <- sapply(strsplit(basename(sample_list), "_"), `[`, 1)
sample_output <- file.path(sample_output_path, paste0(sample_data.names, "_filt.fastq.gz"))
sample_output
names(sample_output) <- sample_data.names
# Running this step will take a couple of minutes to run, we are processing millions of
# sequencing reads. If your CPU has multiple cores, you can try setting `multithread = TRUE`
# in the parameters. Set verbose to true to get updates where we are in processing.
dada_output <- CompHeatmaps::preprocess_16s_data(sample_list, sample_output, verbose = TRUE, multithread = TRUE)
```

Let's make a sequence table to analyze the abundance of our processed reads.
```{r}
# Make a sequence table
seq_table <- create_abundance_table(dada_output)
rownames(seq_table) <- sample_data.names
# View the dimensions, we will have 1 row for each sample and 1 column for each unique read
dim(seq_table)
class(seq_table)
seq_table[502]
```

Finally, we can plot our data using the `create_heatmap` function.
```r
create_heatmap(seq_table)
```


## References

- A framework for human microbiome research. Human Microbiome Project
  Consortium, Nature, 486 (2012), pp. 215-221.

- Structure, function and diversity of the healthy human microbiome.
  Human Microbiome Project Consortium, Nature, 486 (2012), pp. 207–214.

- Warnes G, Bolker B, Bonebakker L, Gentleman R, Huber W, Liaw A, Lumley
  T, Maechler M, Magnusson A, Moeller S, Schwartz M, Venables B (2022).
  *gplots: Various R Programming Tools for Plotting Data*. R package
  version 3.1.3, <https://CRAN.R-project.org/package=gplots>.

- Callahan BJ, McMurdie PJ, Rosen MJ, Han AW, Johnson AJA, Holmes SP
  (2016). “DADA2: High-resolution sample inference from Illumina
  amplicon data.” *Nature Methods*, *13*, 581-583.
  <doi:10.1038/nmeth.3869> <https://doi.org/10.1038/nmeth.3869>.

- phyloseq: An R package for reproducible interactive analysis and
  graphics of microbiome census data. Paul J. McMurdie and Susan Holmes

  2013) PLoS ONE 8(4):e61217.

- Gu, Z. (2014) circlize implements and enhances circular visualization
  in R. Bioinformatics.

- Müller K (2020). *here: A Simpler Way to Find Your Files*. R package
  version 1.0.1, <https://CRAN.R-project.org/package=here>.

- Michał Krassowski. (2020). krassowski/complex-upset. Zenodo.
  <http://doi.org/10.5281/zenodo.3700590>

- Alexander Lex, Nils Gehlenborg, Hendrik Strobelt, Romain Vuillemot,
  Hanspeter Pfister, UpSet: Visualization of Intersecting Sets, IEEE
  Transactions on Visualization and Computer Graphics (InfoVis ’14),
  vol. 20, no. 12, pp. 1983–1992, 2014.

- Wickham H, Hester J, Chang W, Bryan J (2022). *devtools: Tools to Make
  Developing R Packages Easier*. R package version 2.4.5,
  <https://CRAN.R-project.org/package=devtools>.